File: timbre/tools/browse_url.py
from __future__ import annotations
import asyncio
import os
from pathlib import Path
from .registry import Tool, register

MAX_CHARS = 50_000


def _parse_netscape_cookies(path: str) -> list[dict]:
    cookies = []
    try:
        with open(path, encoding="utf-8") as f:
            for line in f:
                line = line.strip()
                if not line or line.startswith("#"):
                    continue
                parts = line.split("\t")
                if len(parts) < 7:
                    continue
                domain, _, path_, secure, expires, name, value = parts[:7]
                cookies.append({
                    "name": name, "value": value,
                    "domain": domain.lstrip("."), "path": path_,
                    "secure": secure.upper() == "TRUE", "sameSite": "None",
                })
    except Exception:
        pass
    return cookies


async def _handler(url: str) -> dict:
    try:
        from playwright.async_api import async_playwright
    except ImportError:
        return {"content": "", "error": "playwright not installed — run: playwright install chromium", "url": url}

    cookies_file = os.getenv("BROWSER_COOKIES_FILE")

    async with async_playwright() as pw:
        browser = await pw.chromium.launch(headless=True)
        ctx = await browser.new_context(
            user_agent=(
                "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) "
                "AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"
            )
        )
        if cookies_file and Path(cookies_file).exists():
            cookies = _parse_netscape_cookies(cookies_file)
            if cookies:
                await ctx.add_cookies(cookies)

        page = await ctx.new_page()
        try:
            await page.goto(url, wait_until="domcontentloaded", timeout=30_000)
            await asyncio.sleep(2)
            text = await page.evaluate("""() => {
                for (const el of document.querySelectorAll(
                    'nav, footer, header, aside, .ad, .ads, .advertisement, .paywall, script, style'
                )) el.remove();
                return document.body?.innerText || '';
            }""")
            return {"content": text[:MAX_CHARS], "error": None, "url": url}
        except Exception as e:
            return {"content": "", "error": str(e), "url": url}
        finally:
            await browser.close()


browse_url = Tool(
    name="browse_url",
    description=(
        "Fetch the full text of a URL using a headless browser. "
        "Use for complete articles, especially paywalled sources when cookies are configured."
    ),
    parameters={
        "type": "object",
        "properties": {
            "url": {"type": "string", "description": "The URL to fetch"},
        },
        "required": ["url"],
    },
    handler=_handler,
)

register(browse_url, synthesis_only=True)

File: tools/browse_url.js
const fs = require('fs');
const path = require('path');

const MAX_CHARS = 50000;
// Optional: path to a Netscape cookies.txt exported from Chrome
const COOKIES_FILE = process.env.BROWSER_COOKIES_FILE;

function parseCookiesTxt(filePath) {
  const lines = fs.readFileSync(filePath, 'utf8').split('\n');
  return lines
    .filter(l => l && !l.startsWith('#'))
    .map(l => {
      const parts = l.split('\t');
      if (parts.length < 7) return null;
      return {
        domain: parts[0].replace(/^\./, ''),
        path: parts[2],
        secure: parts[3] === 'TRUE',
        expires: parseInt(parts[4], 10) || undefined,
        name: parts[5],
        value: parts[6].trim(),
      };
    })
    .filter(Boolean);
}

async function browseUrl(url) {
  let playwright;
  try {
    playwright = require('playwright');
  } catch {
    return {
      content: null,
      error: 'playwright not installed. Run: npm install playwright && npx playwright install chromium',
    };
  }

  const browser = await playwright.chromium.launch({ headless: true });
  try {
    const context = await browser.newContext({
      userAgent: 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
    });

    // Load cookies from file if configured (allows accessing paywalled content)
    if (COOKIES_FILE && fs.existsSync(COOKIES_FILE)) {
      const cookies = parseCookiesTxt(COOKIES_FILE);
      if (cookies.length > 0) await context.addCookies(cookies);
    }

    const page = await context.newPage();
    await page.goto(url, { waitUntil: 'domcontentloaded', timeout: 30000 });
    await page.waitForTimeout(2000);

    const text = await page.evaluate(() => {
      // Remove nav, footer, ads, scripts
      const remove = ['nav', 'footer', 'header', 'script', 'style', 'aside', '[class*="ad"]', '[class*="banner"]', '[class*="popup"]'];
      remove.forEach(sel => document.querySelectorAll(sel).forEach(el => el.remove()));

      // Try to find main article content
      const selectors = ['article', 'main', '[class*="article"]', '[class*="content"]', '[class*="story"]', '.post', '#content'];
      for (const sel of selectors) {
        const el = document.querySelector(sel);
        if (el && el.innerText.length > 500) return el.innerText;
      }
      return document.body.innerText;
    });

    return { content: text.slice(0, MAX_CHARS), error: null, url };
  } catch (err) {
    return { content: null, error: `Browse failed: ${err.message}` };
  } finally {
    await browser.close();
  }
}

module.exports = { browseUrl };
